### PR TITLE
[lit-html][labs/ssr-client] Fix compiled template support

### PR DESCRIPTION
**Pull request by AndrewJakubowicz**
Change is an independent part of https://github.com/lit/lit/pull/3984

## Summary

Fix `isTemplateResult` return type to include `CompiledTemplateResult`s, and then fix locations in `lit` that didn't handle `CompiledTemplate`s. Primarily the `cache` directive.

## Changes

 - Fixes the return type of `isTemplateResult` to return the correct type (as per current implementation): a `TemplateResult` or a `CompiledTemplateResult`.
 - Fix cache directive to work with both uncompiled and compiled templates using a helper to access the template strings array.
 - Made it explicit that `hydration` and `experimental-hydration` do not work with compiled templates by making the check more strict. *Maybe could add a dev mode error?*
- Added compiled template cache directive unit test, and `isTemplateResult` type tests.

## Considerations

I marked this as a minor change, however it is a breaking change in terms of TypeScript types. The return type of `isTemplateResult` without a second argument now has an additional consideration and could break users of `isTemplateResult`. However the type now more truthfully conveys what `isTemplateResult` is asserting.

## Alternative considered:
 - Make `isTemplateResult` stricter such that the type is only a `TemplateResult`. A breaking change (although people should realistically not be depending on `CompiledTemplateResult`s yet)
 - Add the following helpers: `isCompiledTemplateResult` to check for CompiledTemplates, and `getTemplateStringsArray` to generically retrieve the template strings array from either a compiled or uncompiled template result.

## Test plan

Unit tests and type only tests added.

### CODE REVIEW

File: packages/lit-html/src/directive-helpers.ts [hunk 1]
```diff
@@ -42,17 +48,23 @@ export const TemplateResultType = {
...
-  type?: TemplateResultType
-): value is TemplateResult =>
+  type: TemplateResultType | undefined
+) =>
   type === undefined
     ? // This property needs to remain unminified.
       (value as TemplateResult)?.['_$litType$'] !== undefined
-    : (value as TemplateResult)?.['_$litType$'] === type;
+    : (value as TemplateResult)?.['_$litType$'] ===
+      type) as unknown as isTemplateResultType;
```
Comment (rictic) [inline packages/lit-html/src/directive-helpers.ts:68]: I think we want an isCompiledTemplateResult guard function here, otherwise we're repeating the check in a number of places, requiring casts

**Review by rictic — CHANGES_REQUESTED** [review]
Looks good!

File: packages/lit-html/src/directive-helpers.ts [hunk 2]
```diff
@@ -42,17 +48,23 @@ export const TemplateResultType = {
 export type TemplateResultType =
   (typeof TemplateResultType)[keyof typeof TemplateResultType];
 
+type isTemplateResultType = {
+  (val: unknown): val is TemplateResult | CompiledTemplateResult;
+  (val: unknown, type: TemplateResultType): val is TemplateResult;
```
Comment (augustjk) [inline packages/lit-html/src/directive-helpers.ts:53]: This just occurred to me as I was looking through, but wasn't part of the original implementation. Maybe we don't need this kind of granularity, so feel free to ignore.
```suggestion
  <T extends TemplateResultType>(val: unknown, type: T): val is TemplateResult<T>;
```

File: .changeset/chilled-drinks-travel.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+---
+'@lit-labs/ssr-client': patch
+'lit-html': minor
+'lit': minor
+---
+
+Fix return type of `isTemplateResult` helper to include the `CompiledTemplateResult` and fix the `cache` directive to work correctly with `CompiledTemplateResult`s. Add an explicit error message for ssr hydration to throw on `CompiledTemplate`s. Any usage of `isTemplateResult` without a second argument should also be handling the compiled template.
```
Comment (augustjk) [inline .changeset/chilled-drinks-travel.md:7]: I think it would make sense to make 2 separate changeset file, having different lines for `lit-html` and `lit` vs `@lit-labs/ssr-client`.

File: packages/labs/ssr-client/src/lib/hydrate-lit-html.ts [hunk 4]
```diff
@@ -270,24 +270,28 @@ const openChildPart = (
     //   throw new Error('Hydration value mismatch: Primitive found where TemplateResult expected');
     // }
   } else if (isTemplateResult(value)) {
+    if (typeof value['_$litType$'] != 'number') {
```
Comment (augustjk) [inline packages/labs/ssr-client/src/lib/hydrate-lit-html.ts:273]: Nit:
```suggestion
    if (typeof value['_$litType$'] !== 'number') {
```

Comment (augustjk) [reply packages/labs/ssr-client/src/lib/hydrate-lit-html.ts:273]: But also, could use `isCompiledTemplateResult` checker like @rictic suggested.

**Review by augustjk — CHANGES_REQUESTED** [review]
Feel free to push back on any requested changes.

File: packages/labs/ssr-client/src/lib/hydrate-lit-html.ts (hunk 4 above)
Comment (AndrewJakubowicz) [reply packages/labs/ssr-client/src/lib/hydrate-lit-html.ts]: Added `isCompiledTemplateResult` check.

File: .changeset/chilled-drinks-travel.md (hunk 3 above)
Comment (AndrewJakubowicz) [reply .changeset/chilled-drinks-travel.md]: Great suggestion - done

File: packages/lit-html/src/directive-helpers.ts (hunk 1 above)
Comment (AndrewJakubowicz) [reply packages/lit-html/src/directive-helpers.ts:80]: Great! Done

File: packages/lit-html/src/directive-helpers.ts (hunk 2 above)
Comment (AndrewJakubowicz) [reply packages/lit-html/src/directive-helpers.ts]: Awesome!

File: packages/lit-html/src/directive-helpers.ts [hunk 5]
```diff
@@ -42,17 +48,23 @@ export const TemplateResultType = {
...
-  type?: TemplateResultType
-): value is TemplateResult =>
+  type: TemplateResultType | undefined
+) =>
   type === undefined
     ? // This property needs to remain unminified.
       (value as TemplateResult)?.['_$litType$'] !== undefined
-    : (value as TemplateResult)?.['_$litType$'] === type;
+    : (value as TemplateResult)?.['_$litType$'] ===
+      type) as unknown as isTemplateResultType;
```
[the remark this answers is not in the document]
Comment (AndrewJakubowicz) [reply packages/lit-html/src/directive-helpers.ts]: Done

**Review by rictic — APPROVED** [review]

**Review by augustjk — APPROVED** [review]

**Comment by AndrewJakubowicz** [comment]
I added two more small cleanup commits. 1. Use the `isCompiledTemplate` helper in the cache directive fix to simplify some logic. (No behavior change). 2. Add type only tests to verify the excellent improvement in returning the exact (HTML or SVG) TemplateResult type from `isTemplateResult`.

PTAL

File: packages/lit-html/src/directive-helpers.ts [hunk 6]
```diff
@@ -42,17 +48,35 @@ export const TemplateResultType = {
 export type TemplateResultType =
   (typeof TemplateResultType)[keyof typeof TemplateResultType];
 
+type isTemplateResultType = {
```
Comment (justinfagnani) [inline packages/lit-html/src/directive-helpers.ts:51]: Capitalize all type names, and I wouldn't include "Type" in the name. I think `IsTemplateResult` is fine.

File: packages/lit-html/src/directive-helpers.ts [hunk 7]
```diff
@@ -42,17 +48,35 @@ export const TemplateResultType = {
...
-  type?: TemplateResultType
-): value is TemplateResult =>
+  type: TemplateResultType | undefined
+) =>
   type === undefined
     ? // This property needs to remain unminified.
       (value as TemplateResult)?.['_$litType$'] !== undefined
-    : (value as TemplateResult)?.['_$litType$'] === type;
+    : (value as TemplateResult)?.['_$litType$'] ===
+      type) as isTemplateResultType;
```
Comment (justinfagnani) [inline packages/lit-html/src/directive-helpers.ts:70]: Is this `as isTemplateResultType` cast is in the wrong place? It reads to me like it's casting a boolean to a function call interface, but with a body-less arrow function I'm not sure of the precedence here.

To make it more clear I would type the `isTemplateResult` variable, instead of a cast, like:

```ts
export const isTemplateResult: IsTemplateResult = ((
  value: unknown,
  type: TemplateResultType | undefined
) =>
```

File: packages/lit-html/src/directives/cache.ts [hunk 8]
```diff
@@ -21,13 +22,25 @@ import {
   clearPart,
   getCommittedValue,
   insertPart,
+  isCompiledTemplateResult,
   isTemplateResult,
   setCommittedValue,
 } from '../directive-helpers.js';
 
+/**
+ * @returns The template strings array for the provided template.
```
Comment (justinfagnani) [inline packages/lit-html/src/directives/cache.ts:31]: I would comment the string contents are not compatible, since compiled template strings have already been prepared. They're only suitable for use as a key.

File: packages/lit-html/src/directives/cache.ts [hunk 9]
```diff
@@ -21,13 +22,25 @@ import {
...
+ */
+function getStringsFromTemplate(
```
Comment (justinfagnani) [inline packages/lit-html/src/directives/cache.ts:33]: Use an arrow function

File: packages/lit-html/src/directives/cache.ts [hunk 10]
```diff
@@ -21,13 +22,25 @@ import {
...
+  template: TemplateResult | CompiledTemplateResult
```
Comment (justinfagnani) [inline packages/lit-html/src/directives/cache.ts:34]: It's not a template, but a template result. I would name it `result`, and put "result" in the function name too.

File: packages/lit-html/src/directives/cache.ts [hunk 11]
```diff
@@ -45,17 +58,23 @@ class CacheDirective extends Directive {
...
+      let cachedContainerPart = this._templateCache.get(
+        getStringsFromTemplate(this._value)
+      );
       if (cachedContainerPart === undefined) {
         const fragment = document.createDocumentFragment();
         cachedContainerPart = render(nothing, fragment);
         cachedContainerPart.setConnected(false);
-        this._templateCache.set(this._value.strings, cachedContainerPart);
+        this._templateCache.set(
+          getStringsFromTemplate(this._value),
```
Comment (justinfagnani) [inline packages/lit-html/src/directives/cache.ts:75]: You call this several times. Can you put the result in a variable?

File: packages/lit-html/src/directive-helpers.ts (hunk 7 above)
Comment (AndrewJakubowicz) [reply packages/lit-html/src/directive-helpers.ts:70]: Went through this with @rictic and unfortunately the cast is required. If I move the type as you've suggested it fails with: `Type '(value: unknown, type: TemplateResultType | undefined) => boolean' is not assignable to type 'IsTemplateResult'.`

File: packages/lit-html/src/directives/cache.ts (hunk 10 above)
Comment (AndrewJakubowicz) [reply packages/lit-html/src/directives/cache.ts:34]: I.e. `getStringsFromTemplateResult`?

File: packages/lit-html/src/directive-helpers.ts (hunk 7 above)
Comment (augustjk) [reply packages/lit-html/src/directive-helpers.ts:70]: https://github.com/lit/lit/pull/3993#discussion_r1255004607

See the snippet here that moves the type next to the specifier. You need the 2nd param to be optional and also one of the type predicate return annotation to satisfy TS on the implementation code.

File: packages/lit-html/src/directives/cache.ts (hunk 8 above)
Comment (AndrewJakubowicz) [reply packages/lit-html/src/directives/cache.ts:31]: Something like?

```

/**
 * The template strings array contents are not compatible between the two
 * template result types as the compiled template contains a prepared string,
 * only use for cache key.
 * @returns The template strings array for the provided template result.
 */
 ```

File: packages/lit-html/src/directive-helpers.ts (hunk 7 above)
Comment (AndrewJakubowicz) [reply packages/lit-html/src/directive-helpers.ts:70]: Oh - I am so sorry. It did not click that you had removed the assertion!

File: packages/lit-html/src/directives/cache.ts (hunk 11 above)
Comment (AndrewJakubowicz) [reply packages/lit-html/src/directives/cache.ts]: Done. Lifted the calls to the top.

File: packages/lit-html/src/directives/cache.ts (hunk 9 above)
Comment (augustjk) [reply packages/lit-html/src/directives/cache.ts]: @justinfagnani For my own curiosity's sake, is the main reason for preferring arrow function here for the brevity of the implicit return?

**Review by augustjk — APPROVED** [review]

File: packages/lit-html/src/directives/cache.ts (hunk 9 above)
Comment (justinfagnani) [reply packages/lit-html/src/directives/cache.ts]: A few reasons:
* Consistency
* As an indicator that this function doesn't need a bindable `this`, or rather `function` should be an indicator that a function _does_ need a bindable `this`. Similar to preferring `const` so that `let` is an indicator that a variable will be reassigned.
* It's a few less characters, and because of our consistent usage it _may_ compress better in gzip (brotli likely as a pre-baked dictionary entry for `function`, but I'd have to look that up).

**Review by justinfagnani — APPROVED** [review]
